controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,448 @@
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3v1alpha1.BareMetalHost{}
+	if err = r.Get(context.TODO(), req.NamespacedName, bmh); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "could not get baremetalhost")
+	}
+
+	// Fetch the HostFirmwareSettings
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	info := &rInfo{log: reqLogger, hfs: hfs, bmh: bmh}
+	if err = r.Get(ctx, req.NamespacedName, hfs); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// A resource doesn't exist, create one if there is a baremetalhost
+			if err = r.newHostFirmwareSettings(info, req.NamespacedName); err != nil {
+
+				return ctrl.Result{}, errors.Wrap(err, "could not create firmware settings")
+			}
+		} else {
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+		}
+	}
+
+	// Create a provisioner that can access Ironic API
+	provID := bmh.Status.Provisioning.ID
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildEmptyHostData(provID), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+	info.log.Info("created a provisioner for host", "uuid", provID)
+
+	ready, err := prov.IsReady()
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check services availability")
+	}
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	// Get the data from Ironic and update HFS, the call to provisioner may fail if settings can't
+	// be retrieved for the bios_interface. Just log it and don't retry
+	if err = r.updateHostFirmwareSettings(prov, info); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Could not update hostFirmwareSettings")
+	}
+
+	for _, e := range info.events {
+		r.publishEvent(req, e)
+	}
+
+	// only run again on changes
+	return ctrl.Result{}, nil
+}
+
+// Get the firmware settings from the provisioner and update hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) updateHostFirmwareSettings(prov provisioner.Provisioner, info *rInfo) (err error) {
+
+	info.log.Info("retrieving firmware settings and saving to resource")
```

Comment:
nit: Maybe add BMH and/or Node UUID?

Reply:
Good point, added node uuid here so can remove log below